core/src/main/java/org/apache/james/core/User.java

```diff
@@ -99,11 +99,7 @@ public User withDefaultDomain(Domain defaultDomain) {
     }
 
     public User withDefaultDomainFromUser(User other) {
-        Preconditions.checkArgument(other.hasDomainPart());
-        if (hasDomainPart()) {
-            return this;
-        }
-        return new User(localPart, Optional.of(other.domainPart.get()));
+        return withDefaultDomain(other.domainPart.get());
```

Comment:
this is a public API for User class. You cannot assume other will have a domain, and we can considerate as legit for other to not have a domain. I think you should you get domain from other (empty or non empty) without wondering if it has a domain